SwiftCompilerSources/Sources/Optimizer/FunctionPasses/AutodiffClosureSpecialization.swift

```diff
@@ -0,0 +1,180 @@
+///     %2 = apply %1(%0): $Float
+///     return %2: $Float
+/// }
+/// 
+/// sil @vjp_foo: $(Float) -> (Float, (Float) -> Float) { 
+/// bb0(%0: $Float):
+///     %1 = apply @sin(%0): $Float
+///     %2 = partial_apply @pb_sin(%0): $(Float) -> Float
+///     %pb_foo = partial_apply @pb_foo(%2): $(Float, (Float) -> Float) -> Float
+///     return (%y3, %pb_foo)
+/// }
+/// 
+/// //============== After optimization ==============//
+/// sil @specialized_pb_foo: $(Float, Float) -> Float { 
+/// bb0(%0: $Float, %1: $Float):
+///     %2 = partial_apply @pb_sin(%1): $(Float) -> Float
+///     %3 = apply %2(%0): $Float
+///     return %3: $Float
+/// }
+/// 
+/// sil @vjp_foo: $(Float) -> (Float, (Float) -> Float) { 
+/// bb0(%0: $Float):
+///     %1 = apply @sin(%0): $Float
+///     %pb_foo = partial_apply @specialized_pb_foo(%0): $(Float, Float) -> Float
+///     return (%y3, %pb_foo)
+/// }
+/// ``` 
+///
+/// Similar complications, as in the existing closure-specialization optimization, exist
+/// when populating the specialized Pullback and rewriting the apply instruction in the 
+/// VJP. The details of the complications and our work arounds are listed below 
+/// (copied over from the existing closure-specialization pass):
+///
+/// 1. If we support the specialization of closures with multiple user callsites
+///    that can be specialized, we need to ensure that any captured values have
+///    their reference counts adjusted properly. This implies for every
+///    specialized call site, we insert an additional retain for each captured
+///    argument with reference semantics. We will pass them in as extra @owned
+///    to the specialized function. This @owned will be consumed by the "copy"
+///    partial apply that is in the specialized function. Now the partial apply
+///    will own those ref counts. This is unapplicable to thin_to_thick_function
+///    since they do not have any captured args.
+///
+/// 2. If the closure was passed in @owned vs if the closure was passed in
+///    @guaranteed. If the original closure was passed in @owned, then we know
+///    that there is a balancing release for the new "copy" partial apply. But
+///    since the original partial apply no longer will have that corresponding
+///    -1, we need to insert a release for the old partial apply. We do this
+///    right after the old call site where the original partial apply was
+///    called. This ensures we do not shrink the lifetime of the old partial
+///    apply. In the case where the old partial_apply was passed in at +0, we
+///    know that the old partial_apply does not need to have any ref count
+///    adjustments. On the other hand, the new "copy" partial apply in the
+///    specialized function now needs to be balanced lest we leak. Thus we
+///    insert a release right before any exit from the function. This ensures
+///    that the release occurs in the epilog after any retains associated with
+///    @owned return values.
```

Comment:
All this is much easier to handle in OSSA

Reply:
Same as previous comment.